src/HttpOverStream.Client/DialMessageHandler.cs

```diff
@@ -62,42 +63,78 @@ protected override Task<Stream> CreateContentReadStreamAsync()
         protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             ValidateAndNormalizeRequest(request);
-            var stream = await _dial.DialAsync(request, cancellationToken).ConfigureAwait(false);
-
-            request.Properties.Add(UnderlyingStreamProperty, stream);
-            await stream.WriteMethodAndHeadersAsync(request, cancellationToken).ConfigureAwait(false);
-            if (request.Content != null)
+            Stream stream = null;
+            try
             {
-                await request.Content.CopyToAsync(stream).ConfigureAwait(false);
-            }
-            await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                Debug.WriteLine("Client: About to connect");
+                stream = await _dial.DialAsync(request, cancellationToken).ConfigureAwait(false);
 
-            var responseContent = new DialResponseContent();
-            var response = new HttpResponseMessage { RequestMessage = request, Content = responseContent };
-            string statusLine = await stream.ReadLineAsync(cancellationToken).ConfigureAwait(false);
-            ParseStatusLine(response, statusLine);
-            for (; ; )
-            {
-                var line = await stream.ReadLineAsync(cancellationToken).ConfigureAwait(false);
-                if (line.Length == 0)
+                Debug.WriteLine("Client: Connected");
+                request.Properties.Add(UnderlyingStreamProperty, stream);
+
+                Debug.WriteLine("Client: Writing request");
+                await stream.WriteMethodAndHeadersAsync(request, cancellationToken).ConfigureAwait(false);
+
+                // as soon as headers are sent, we should begin reading the response, and send the request body concurrently
+                // This is because if the server 404s nothing will ever read the response and it'll hang waiting
+                // for someone to read it
+
+                var writeContentCTS = new CancellationTokenSource();
+                var writeContentToken = writeContentCTS.Token;
+                cancellationToken.Register(() => writeContentCTS.Cancel());

```

Comment:
we dont actually have access to the original cancellationTokenSource here which is why I didn't do this, and I'm not keen on passing the source around because it means anything can cancel the token, but maybe thats fine?

Reply:
CancellationTokenSource.CreateLinkedTokenSource() takes cancellation tokens (not sources) as parameters, and does the equivalent in the code here in a less verbose, more optimized way.